errors: fix the formatting with %+v

knz opened the pull request:
(found by @RaduBerinde; needed to complete #38570)

The new library `github.com/cockroachdb/errors` was not implementing
`%+v` formatting properly for assertion and unimplemented errors.
The wrong implementation was hiding the details of the cause
of these two error types from the formatting logic.

Fixing this bug comprehensively required completing the investigation
of the Go 2 / `xerrors` error proposal. This revealed that the
implementation of `fmt.Formatter` for wrapper errors (a `Format()`
method) is required in all cases, at least until Go's stdlib
learns about `errors.Formatter`. More details at
https://github.com/golang/go/issues/29934 and this commit message: https://github.com/cockroachdb/errors/commit/78b6caa8a0455b5fa93f17d4b4c3132597342c0e.

This patch bumps the dependency `github.com/cockroachdb/errors` to
pick up the fixes to assertion failures and unimplemented errors.

The new definition of `errors.FormatError()` subsequently required
re-implemening `Format)` for `pgerros.withCandidateCode`, which is
also done here.

Finally, this patch also picks up `errors.As()` and the new
streamlined `fmt.Formatter` / `errors.Formatter` interaction, so this
patch also simplifies a few custom error types in CockroachDB
accordingly.

Release note: None

knz committed:
errors: fix the formatting with %+v

(found by @RaduBerinde)

The new library `github.com/cockroachdb/errors` was not implementing
`%+v` formatting properly for assertion and unimplemented errors.
The wrong implementation was hiding the details of the cause
of these two error types from the formatting logic.

Fixing this bug comprehensively required completing the investigation
of the Go 2 / `xerrors` error proposal. This revealed that the
implementation of `fmt.Formatter` for wrapper errors (a `Format()`
method) is required in all cases, at least until Go's stdlib
learns about `errors.Formatter`. More details at
https://github.com/golang/go/issues/29934.

This patch bumps the dependency `github.com/cockroachdb/errors` to
pick up the fixes to assertion failures and unimplemented errors.

The new definition of `errors.FormatError()` subsequently required
re-implemening `Format)` for `pgerros.withCandidateCode`, which is
also done here.

Finally, this patch also picks up `errors.As()` and the new
streamlined `fmt.Formatter` / `errors.Formatter` interaction, so this
patch also simplifies a few custom error types in CockroachDB
accordingly.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38710)
<!-- Reviewable:end -->

knz committed:
*: avoid using sub-packages of cockroachdb/errors

Since the APIs are exported in the top level `errors`,
it is better for dev UX to have a single package to remember.

Release note: None

RaduBerinde reviewed (commented):
LGTM. Thanks for the fix! I verified that the stack trace shows up if I wrap runtime errors in assertion errors and use `%+v` in the opt tester.

In the future, I think it would be good to post a PR for the `errors` change (in that repo) so that code gets reviewed; it was a non-trivial change (btw I think `FormatError` could use some unit tests).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38710#-:-LjEQBgG52v7D4E-grWM:b-9fof78)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

knz commented:
I agree! and I'll welcome your reviews.

That being said FormatError does get tested, extensively. There are two dozens unit tests in package errbase then a dozen more spread out through the other sub-packages.

knz commented:
bors r+
